src/verbs/queue_pair.rs

```diff
@@ -38,26 +42,220 @@ pub enum QueuePairState {
     Unknown = ibv_qp_state::IBV_QPS_UNKNOWN,
 }
 
+impl From<u32> for QueuePairState {
+    fn from(state: u32) -> Self {
+        match state {
+            ibv_qp_state::IBV_QPS_RESET => QueuePairState::Reset,
+            ibv_qp_state::IBV_QPS_INIT => QueuePairState::Init,
+            ibv_qp_state::IBV_QPS_RTR => QueuePairState::ReadyToReceive,
+            ibv_qp_state::IBV_QPS_RTS => QueuePairState::ReadyToSend,
+            ibv_qp_state::IBV_QPS_SQD => QueuePairState::SendQueueDrain,
+            ibv_qp_state::IBV_QPS_SQE => QueuePairState::SendQueueError,
+            ibv_qp_state::IBV_QPS_ERR => QueuePairState::Error,
+            ibv_qp_state::IBV_QPS_UNKNOWN => QueuePairState::Unknown,
+            _ => panic!("Unknown qp state: {state}"),
+        }
+    }
+}
+
+pub trait QueuePair {
+    //! return the basic handle of QP;
+    //! we mark this method unsafe because the lifetime of ibv_qp is not
+    //! associated with the return value.
+    unsafe fn qp(&self) -> NonNull<ibv_qp>;
+
+    fn modify(&mut self, attr: &QueuePairAttribute) -> Result<(), String> {
+        // ibv_qp_attr does not impl Clone trait, so we use struct update syntax here
+        let mut qp_attr = ibv_qp_attr { ..attr.attr };
+        let ret = unsafe { ibv_modify_qp(self.qp().as_ptr(), &mut qp_attr as *mut _, attr.attr_mask.bits) };
+        if ret == 0 {
+            Ok(())
+        } else {
+            check_transition_mask(attr.attr_mask, self.state(), attr.attr.qp_state.into()).unwrap();
```

Comment:
`attr.attr.qp_state` may be invalid since the user may not set the `qp_state` in each modify_qp operation.

Reply:
I would check `mask & IBV_QP_STATE` first, and decide whether use `attr.qp_state` or not